src/builder.rs

```diff
@@ -0,0 +1,54 @@
+use crate::{errors::Error, Layer};
+
+#[derive(Debug)]
+pub struct PacketBuilder {
+    layers: Vec<Box<dyn Layer + Send>>,
+    raw_bytes: Option<Vec<u8>>,
+}
+
+impl PacketBuilder {
+    pub fn new() -> Self {
+        PacketBuilder {
+            layers: vec![],
+            raw_bytes: None,
+        }
+    }
+
+    pub fn stack(mut self, layer: Box<dyn Layer + Send>) -> Result<Self, Error> {
+        if self.raw_bytes.is_none() {
+            self.layers.push(layer);
+            Ok(self)
+        } else {
+            Err(Error::SculptingError(format!(
+                "Cannot push layer on top of raw byte layer"
+            )))
+        }
+    }
+
+    pub fn stack_bytes(mut self, bytes: Vec<u8>) -> Self {
+        self.raw_bytes = Some(bytes);
+        self
+    }
+
+    pub fn build(&self) -> Result<Vec<Vec<u8>>, Error> {
+        let mut result = vec![];
+
+        if let Some(bytes) = &self.raw_bytes {
+            result.push(bytes.clone());
+        }
+
+        for i in (0..=self.layers.len() - 1).rev() {
+            // Layer following this is either raw bytes or empty
+            if i + 1 >= self.layers.len() {
+                result.push(self.layers[i].encode_bytes(result.last(), None)?);
+            } else {
+                result.push(
+                    self.layers[i].encode_bytes(result.last(), Some(self.layers[i + 1].name()))?,
+                );
+            }
```

Comment:
Make this more idiomatic.

Reply:
I believe the complexity of this logic can be pushed into `stack` as described above and then `build` will be a trivial function.

Also, once *correct* `Packet` structure is built, `to_bytes` would be quite simple. 

do a per layer `to_bytes` and then finally `fixup_checkums` (using some metadata of the packet).